noodles-fasta/src/fai/indexer.rs

```diff
@@ -0,0 +1,247 @@
+        Self { inner, offset: 0 }
+    }
+
+    /// Reads a single sequence line.
+    ///
+    /// Trailing whitespaces are not modified.
+    ///
+    /// Returns the number of bytes read from the stream. If the position of the stream is at
+    /// a definition or EOF has been reached, the number of bytes read is 0.
+    fn read_sequence_line(&mut self, buf: &mut Vec<u8>) -> io::Result<usize> {
+        let mut bytes_read = 0;
+        let mut end_line = false;
+
+        loop {
+            let reader_buf = self.inner.fill_buf()?;
+
+            if end_line || reader_buf.is_empty() || reader_buf[0] == DEFINITION_PREFIX {
+                break;
+            }
+
+            let len = match memchr(NEWLINE, reader_buf) {
+                Some(i) => {
+                    // Do not consume newline.
+                    buf.extend(&reader_buf[..=i]);
+                    end_line = true;
+                    i + 1
+                }
+                None => {
+                    buf.extend(reader_buf);
+                    reader_buf.len()
+                }
+            };
+
+            self.inner.consume(len);
+
+            bytes_read += len;
+        }
+
+        Ok(bytes_read)
+    }
+
+    /// Creates a single index [`Record`](struct.Record.html).
+    ///
+    /// Returns `None` if EOF has been reached. Will return an error if:
+    /// * Stream is not at the start of a definition
+    /// * No sequence after a definition
+    /// * Sequence lines are not the same length (excluding the last line)
+    pub fn index_record(&mut self) -> Result<Option<Record>, IndexError> {
+        let mut def_str = String::new();
+        let mut line = Vec::new();
+        let mut length = 0u64;
+
+        let def_len = self.inner.read_line(&mut def_str)?;
+        if def_len == 0 {
+            return Ok(None);
+        }
+
+        self.offset += def_len as u64;
+        let index_offset = self.offset;
+        let def: Definition = def_str.trim_end().parse()?;
+
+        // The first sequence line determines how long each line should be.
+        let mut seq_len = self.read_sequence_line(&mut line)? as u64;
+        let line_width = seq_len;
+        let line_bases = len_with_right_trim(&line) as u64;
+
+        loop {
+            let prev_line_width = seq_len;
+            let prev_line_bases = len_with_right_trim(&line) as u64;
+            length += prev_line_bases;
+            self.offset += prev_line_width;
+
+            line.clear();
+            seq_len = self.read_sequence_line(&mut line)? as u64;
+            if seq_len == 0 {
+                break;
+            // If there are more lines, check the previous line has equal length to first.
+            } else if prev_line_width != line_width || prev_line_bases != line_bases {
+                return Err(IndexError::InvalidLineLength(self.offset));
+            }
+        }
+
+        if length == 0 {
+            return Err(IndexError::EmptyDefinition(self.offset));
```

Comment:
If `length` is 0, doesn't that mean the sequence is empty and not the definition?